### PR TITLE
net: buf: add function to compare data with a net_buf's content

### PR DESCRIPTION
**Pull request by kderda**
This PR introduces:
- new function for comparing data under the pointer with a net_buf's content (or multiple buffers when needed), it feels like a nice improvement to be able to match some bytes without a need of temporary buffers
- unit test for the provided change

### CODE REVIEW

File: include/zephyr/net/buf.h [hunk 1]
```diff
@@ -2414,6 +2414,23 @@ size_t net_buf_append_bytes(struct net_buf *buf, size_t len,
...
+ * @param buf Network buffer
+ * @param offset Starting offset to compare from
+ * @param data Data buffer for comparison
+ * @param data_len Length of the data buffer
+ * @param len Number of bytes to compare
+ *
+ * @return true if the data matches, false otherwise
+ */
+bool net_buf_data_cmp(struct net_buf *buf, size_t offset, void *data,
+		      size_t data_len, size_t len);
```
Comment (pdgendt) [inline include/zephyr/net/buf.h:2432]: Traditional "compare" functions return <0, 0 or >0, maybe rename it to `net_buf_data_eq` or convert the return type to `int` based on the first different byte found (similar to `strcmp`).

Comment (jukkar) [reply include/zephyr/net/buf.h:2432]: I agree, it might be useful to know how many bytes matched.

Comment (konradderda) [reply include/zephyr/net/buf.h]: @pdgendt @jukkar The function now returns a number of bytes matching.

File: include/zephyr/net/buf.h [hunk 2]
```diff
@@ -2414,6 +2414,24 @@ size_t net_buf_append_bytes(struct net_buf *buf, size_t len,
...
+ * @param buf Network buffer
+ * @param offset Starting offset to compare from
+ * @param data Data buffer for comparison
+ * @param data_len Length of the data buffer
+ * @param len Number of bytes to compare
+ *
+ * @return The number of bytes compared before the first difference.
+ */
+size_t net_buf_data_cmp(struct net_buf *buf, size_t offset, void *data,
+		      size_t data_len, size_t len);
```
Comment (pdgendt) [inline include/zephyr/net/buf.h:2433]: Nit: Alignment

Comment (konradderda) [reply include/zephyr/net/buf.h]: Fixed

File: include/zephyr/net/buf.h [hunk 3]
```diff
@@ -2414,6 +2414,24 @@ size_t net_buf_append_bytes(struct net_buf *buf, size_t len,
...
+ * @param buf Network buffer
+ * @param offset Starting offset to compare from
+ * @param data Data buffer for comparison
+ * @param data_len Length of the data buffer
+ * @param len Number of bytes to compare
+ *
+ * @return The number of bytes compared before the first difference.
+ */
+size_t net_buf_data_cmp(struct net_buf *buf, size_t offset, void *data,
+			size_t data_len, size_t len);
```
Comment (jhedberg) [inline include/zephyr/net/buf.h:2433]: Couldn't you just remove the `data_len` parameter? That seems unnecessary. The only place you use it is in the beginning of the function to test `(len > data_len)`. Ensuring that actual valid data behind the `data` pointer is between `0` and `len` should be the responsibility of the caller.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: include/zephyr/net/buf.h (hunk 3 above)
Comment (konradderda) [reply include/zephyr/net/buf.h]: @jhedberg I have removed the parameter

**Review by pdgendt — DISMISSED** [review]
A non-blocking comment:
The `_cmp` suffix and description stating "compare" feels a bit odd, as compare functions usually return a negative/zero/positive result. For example to use in sorting.

Here however, we are finding the longest matching string.

**Review by pdgendt — DISMISSED** [review]
Thanks!

File: tests/net/buf/src/main.c [hunk 4]
```diff
@@ -724,4 +725,60 @@ ZTEST(net_buf_tests, test_net_buf_user_data)
...
+	}
+
+	/* Allocate a single net_buf  */
+	buf = net_buf_alloc(&fixed_pool, K_NO_WAIT);
+	zassert_not_null(buf, "Failed to get buffer");
+
+	written = net_buf_append_bytes(buf, buf->size, data, K_NO_WAIT, NULL, NULL);
+	zassert_equal(written, buf->size, "Failed to fill the buffer");
+	zassert_equal(buf->frags, NULL, "Additional buffer allocated");
+
```
Comment (jukkar) [inline tests/net/buf/src/main.c:749]: Could we also add a test that should fail (negative testing), now all the tests verify that the function passes ok.

File: subsys/net/buf.c [hunk 5]
```diff
@@ -694,3 +694,48 @@ size_t net_buf_append_bytes(struct net_buf *buf, size_t len,
...
+		return 0;
+	}
+
+	left = len;
+	bptr = buf->data + offset;
+
+	while (buf && left > 0) {
+		to_compare = MIN(left, buf->len - offset);
+
+		for (int i = 0; i < to_compare; ++i) {
```
[the remark this answers is not in the document]
Comment (konradderda) [reply subsys/net/buf.c]: Done

File: subsys/net/buf.c [hunk 6]
```diff
@@ -694,3 +694,48 @@ size_t net_buf_append_bytes(struct net_buf *buf, size_t len,
...
+		offset -= buf->len;
+		buf = buf->frags;
+	}
+
+	/* check for out of bounds offset */
+	if (!buf) {
+		return 0;
+	}
+
+	left = len;
```
[the remark this answers is not in the document]
Comment (konradderda) [reply subsys/net/buf.c]: Done

**Review by pdgendt — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**